### PR TITLE
Allow arbitrarily serializable content within `rememberSaveable`

### LINKED ISSUE #1384: Allow arbitrarily serializable content within `rememberSaveable`
Currently, `rememberSaveable` only allows a `MutableState<T>`, where `T` is one of the `JsonPrimitive` (`String`, `Boolean`, `Number`), `List<JsonPrimitive>`, or `Map<JsonPrimitive, JsonPrimitive>`. If you want to serialize anything beyond that, we currently need to add explicit support for it in `StateSnapshot.kt`, and if it's a custom type, you're out of luck.

We should instead use the serializers provided by the user via their `Json` instance, so that the user can serialize anything they can write a serializer for.

```kotlin
Json {
  serializersModule = SerializersModule {
    contextual(CustomSerializer)
  }
}

object CustomSerializer : KSerializer<Custom> { … }

class Custom(val value: Int) { … }

rememberSaveable { Custom(42) }
```

### PR DESCRIPTION
**Pull request by veyndan**
Closes #1384.

### CODE REVIEW

File: samples/emoji-search/launcher/src/commonMain/kotlin/com/example/redwood/emojisearch/launcher/EmojiSearchAppSpec.kt [hunk 1]
```diff
@@ -19,13 +19,15 @@ import app.cash.redwood.treehouse.TreehouseApp
...
 import com.example.redwood.emojisearch.treehouse.HostApi
+import com.example.redwood.emojisearch.treehouse.emojiSearchSerializersModule
 import kotlinx.coroutines.flow.Flow
 
 class EmojiSearchAppSpec(
   override val manifestUrl: Flow<String>,
   private val hostApi: HostApi,
 ) : TreehouseApp.Spec<EmojiSearchPresenter>() {
   override val name = "emoji-search"
+  override val serializersModule = emojiSearchSerializersModule
```
Comment (veyndan) [inline samples/emoji-search/launcher/src/commonMain/kotlin/com/example/redwood/emojisearch/launcher/EmojiSearchAppSpec.kt:30]: It took a while to notice that the `serializersModule` had to be set in _two_ places. I remembered that `Zipline.get()` took a `serializersModule`, but didn't realise that the `TreehouseApp.Spec` also needed one.

The stacktrace was very unhelpful, as it just said that there is no serializer attached for some type. This led to me fiddling with how the serializers are set up, without realising that I needed to attach _another_ `serializersModule` somewhere. I eventually figured it out by doing a global search of `EmptySerializersModule` and found out that I can override the `serializersModule` of `TreehouseApp.Spec`. So I ultimately figured it out by happenstance. 

Is there a way to merge the two ways of setting a `serializersModule`, so we only set the `serializersModule` in a single place? (In the case of Emoji Search, it needed to be set in _three_ places!)

Comment (swankjesse) [reply samples/emoji-search/launcher/src/commonMain/kotlin/com/example/redwood/emojisearch/launcher/EmojiSearchAppSpec.kt:30]: I think the best thing is to define a serializersModule instance somewhere in your app’s commonMain, then to use that shared instance everywhere Treehouse is initialized.

We can’t really share the initialization code as it’s not in commonMain.

Comment (veyndan) [reply samples/emoji-search/launcher/src/commonMain/kotlin/com/example/redwood/emojisearch/launcher/EmojiSearchAppSpec.kt:30]: > I think the best thing is to define a serializersModule instance somewhere in your app’s commonMain, then to use that shared instance everywhere Treehouse is initialized.

Yep, that's what I did with `emojiSearchSerializersModule`. The `serializersModule` is defined once and referenced where needed.

> We can’t really share the initialization code as it’s not in commonMain.

👍

File: redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt [hunk 2]
```diff
@@ -18,75 +18,69 @@ package app.cash.redwood.treehouse
...
+    @Suppress("UNCHECKED_CAST")
+    when (value) {
+      is List<*> -> ListSerializer(PolymorphicSerializer(Any::class)) as SerializationStrategy<Any>
+      is MutableState<*> -> MutableStateSerializer as SerializationStrategy<Any>
+      else -> null
+    }
+  }
+  polymorphicDefaultDeserializer(Any::class) { className ->
+    when (className) {
+      "kotlin.collections.ArrayList" -> ListSerializer(PolymorphicSerializer(Any::class))
```
Comment (swankjesse) [inline redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt:63]: What happens to List subtypes other than ArrayList?

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt [hunk 3]
```diff
@@ -18,66 +18,55 @@ package app.cash.redwood.treehouse
...
+        "key3" to listOf(mutableStateOf("str")),
+        "key4" to listOf("str"),
+        "key5" to listOf(null),
+        "key6" to listOf(true),
+        "key7" to listOf(1.0),
+        "key8" to listOf(1.0f),
+        "key9" to listOf(listOf(1, "str")),
+      ),
+    )
+    val serialized = json.encodeToString(stateSnapshot)
```
Comment (swankjesse) [inline redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:50]: Please also assert on the contents of this string? As-is if there’s a backwards-incompatible change in how a string is encoded we won’t get a failing test.

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt [hunk 4]
```diff
@@ -18,66 +18,55 @@ package app.cash.redwood.treehouse
...
+  fun stateSnapshotSerializeThenDeserialize() {
+    val json = Json {
+      useArrayPolymorphism = true
+      serializersModule = SaveableStateSerializersModule
+    }
+    val stateSnapshot = StateSnapshot(
+      mapOf(
+        "key1" to listOf(mutableStateOf(1)),
+        "key2" to listOf(1),
+        "key3" to listOf(mutableStateOf("str")),
```
Comment (swankjesse) [inline redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:42]: Do mutable state values support all this polymorphism?

Comment (swankjesse) [reply redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:42]: Mind adding a test for that?

**Review by swankjesse — APPROVED** [review]

File: redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt (hunk 2 above)
Comment (veyndan) [reply redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt:63]: It'll crash! I'm grouping other List subtypes as part of the "Add support for rest of built-ins serializers." TODO comment above this.

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt (hunk 4 above)
Comment (veyndan) [reply redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:42]: It does and this is the test for that! The `Int` and `String` values are both resolved polymorphically.

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt (hunk 3 above)
Comment (veyndan) [reply redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:51]: For future readers, I tried asserting on `encodeToString`, but the results are different depending on the target its run on (JS gives a different JSON string than the JVM). I'm gonna assert on the `JsonElement` instead.

**Pull request merged**